packages/keyring-controller/src/KeyringController.ts

```diff
@@ -2371,13 +2472,15 @@ export class KeyringController extends BaseController<
     return this.#withControllerLock(async ({ releaseLock }) => {
       const currentSerializedKeyrings = await this.#getSerializedKeyrings();
       const currentPassword = this.#password;
+      const currentKeyringsMetadata = this.#keyringsMetadata;
 
       try {
         return await callback({ releaseLock });
       } catch (e) {
         // Keyrings and password are restored to their previous state
         await this.#restoreSerializedKeyrings(currentSerializedKeyrings);
         this.#password = currentPassword;
+        this.#keyringsMetadata = currentKeyringsMetadata;
```

Comment:
I think that this will not effectively rollback `this.#keyringsMetadata` to its previous state because we are getting a shallow copy of the array (i.e. the underlying referenced object stored at `currentKeyringsMetadata` will be the same as `this.#keyringsMetadata`), so if the array is modified while `callback({ releaseLock }) is executed and an error is thrown, `this.#keyringsMetadata = currentKeyringsMetadata` will have no effect

Reply:
we could use:
```typescript
      const currenctKeyringsMetadata = this.#keyringsMetadata.slice()
```

Reply:
also, there should probably be a test case for this scenario because an error thrown in `callback` could misalign the two arrays and brick the wallet